### DISCUSSION TITLE
Infix methods

### DISCUSSION THREAD

**Discussion by Adiras** [dotnet/csharplang#8192] [Language Ideas]
Consider adding infix notation for method calls in the language syntax, without requiring the use of periods (.) and parentheses (). This enhancement would allow for more natural language-like code and facilitate the creation of domain-specific languages (DSLs) on top of C#.

A similar functionality already exists in Kotin language: https://kotlinlang.org/docs/functions.html#infix-notation.

### Example

The following examples demonstrates the use of the described functionality.

**Old way:**
```csharp
class Player
{
    public void Attack(Enemy enemy) { }
}

class Enemy;
    
Player warrior = new();
Enemy zombie = new();
        
warrior.Attack(zombie);
```

**Using infix methods:**
```csharp
class Player
{
    public infix void Attack(Enemy enemy) { }
}

class Enemy;
    
Player warrior = new();
Enemy zombie = new();
        
warrior Attack zombie;
```

### Language-Level Support

- Extend the C# compiler to recognize infix notation and generate appropriate IL code.
- Ensure that infix methods follow specific rules (e.g., taking exactly one parameter).

### Backward Compatibility

- Existing code should continue to work without modification (wouldn’t automatically benefit from infix notation).
- Developers can opt in to use infix notation where needed (by adding a special keyword to a method).

**Comment by KennethHoff** [comment]
This feels very awkward to parse and I don't really see the benefit to be honest, but I've never seen the suggestion before - cool

**Comment by colejohnson66** [comment]
F# has support for pipes, which is similar to what you want.

https://theburningmonk.com/2011/09/fsharp-pipe-forward-and-pipe-backward/

**Comment by Richiban** [comment]
It should be worth noting that Scala has this, and I'm not sure it's well-liked: https://docs.scala-lang.org/style/method-invocation.html
